prow/plank/reconciler.go

```diff
@@ -0,0 +1,645 @@
+	}
+
+	originalPJ := pj.DeepCopy()
+	pj.SetComplete()
+	return r.pjClient.Patch(r.ctx, pj, ctrlruntimeclient.MergeFrom(originalPJ))
+}
+
+func (r *reconciler) pod(pj *prowv1.ProwJob) (*corev1.Pod, bool, error) {
+	buildClient, buildClientExists := r.buildClients[pj.ClusterAlias()]
+	if !buildClientExists {
+		// TODO: Use terminal error type to prevent requeuing, this wont be fixed without
+		// a restart
+		return nil, false, fmt.Errorf("no build client found for cluster %q", pj.ClusterAlias())
+	}
+
+	pod := &corev1.Pod{}
+	name := types.NamespacedName{
+		Namespace: r.config().PodNamespace,
+		Name:      pj.Name,
+	}
+
+	if err := buildClient.Get(r.ctx, name, pod); err != nil {
+		if kerrors.IsNotFound(err) {
+			return nil, false, nil
+		}
+		return nil, false, fmt.Errorf("failed to get pod: %v", err)
+	}
+
+	return pod, true, nil
+}
+
+func (r *reconciler) startPod(pj *prowv1.ProwJob) (string, string, error) {
+	buildID, err := r.getBuildID(pj.Spec.Job)
+	if err != nil {
+		return "", "", fmt.Errorf("error getting build ID: %v", err)
+	}
+
+	pod, err := decorate.ProwJobToPod(*pj, buildID)
+	if err != nil {
+		return "", "", err
+	}
+	pod.Namespace = r.config().PodNamespace
+
+	client, ok := r.buildClients[pj.ClusterAlias()]
+	if !ok {
+		// TODO: Terminal error to prevent requeuing
+		return "", "", fmt.Errorf("unknown cluster alias %q", pj.ClusterAlias())
+	}
+	err = client.Create(r.ctx, pod)
+	r.log.WithFields(pjutil.ProwJobFields(pj)).Debug("Create Pod.")
+	if err != nil {
+		return "", "", err
+	}
+	return buildID, pod.Name, nil
+}
+
+func (r *reconciler) getBuildID(name string) (string, error) {
+	return pjutil.GetBuildID(name, r.totURL)
+}
+
+// canExecuteConcurrently determines if the cocurrency settings allow our job
+// to be started. We start jobs with a limited concurrency in order, oldest
+// first. This allows us to get away without any global locking by just looking
+// at the jobs in the cluster.
+func (r *reconciler) canExecuteConcurrently(pj *prowv1.ProwJob) (bool, error) {
+
+	if max := r.config().Plank.MaxConcurrency; max > 0 {
+		pjs := &prowv1.ProwJobList{}
+		if err := r.pjClient.List(r.ctx, pjs, optNotCompletedProwJobs()); err != nil {
+			return false, fmt.Errorf("failed to list prowjobs: %w", err)
+		}
+		// The list contains our own ProwJob
+		running := len(pjs.Items) - 1
+		if running >= max {
+			r.log.WithFields(pjutil.ProwJobFields(pj)).Infof("Not starting another job, already %d running.", running)
+			return false, nil
+		}
+	}
+
+	if pj.Spec.MaxConcurrency == 0 {
+		return true, nil
+	}
+
+	pjs := &prowv1.ProwJobList{}
+	if err := r.pjClient.List(r.ctx, pjs, optNotCompletedProwJobsNamed(pj.Spec.Job)); err != nil {
+		return false, fmt.Errorf("failed listing prowjobs: %w:", err)
+	}
+	r.log.Infof("got %d not completed with same name", len(pjs.Items))
+
+	var olderMatchingPJs int
+	for _, foundPJ := range pjs.Items {
+		// Ignore self here. Second half of the condition is needed for tests.
```

Comment:
Production code that has call-outs for tests?!

Reply:
Yeah its terrible and I am embarrassed by this. But not sure how to else test this in a unit test:
* Including the actual Reconciliation in the test makes it more complicated
* I don't want to have a package-level variable for the Reconiler
* I don't want to make passing in the Reconciler part of the public api for this package

Reply:
I'm having a little bit of a hard time wrapping my brain around what specifically you are trying to do with this in testing, mind linking to the diff? GitHub rendering on this PR is really haywire.

Reply:
Wait, my comment was for a different place of the code, I didn't look at the diff. So the basic problem here is:
* In old plank we always reconcile all prowjobs at once
* This allows us to track the pending prowjobs in memory and invalidate it on the end of the reconciliation
* Doing this here would introduce a global lock which I didn't want
* So we look at the jobs in the cluster instead and list all of them (via an index, else it would be incredible expensive)
* This list we get includes ourselves, so we have to remove ourselves via the UID
* The fakeclient does not set a UID

But you are right, the simpler and better approach is to just set a UID on the prowjob under test, I've done that.

The corresponding test is `TestMaxConcurency`